### PR TITLE
Use uuid5() as an option in generate_oid if uuid3 fails.

### PR DESCRIPTION
**Pull request by wenottingham**
uuid3() uses md5, which is disallowed on systems running in FIPS mode
(http://csrc.nist.gov/groups/STM/cavp/validation.html)

### CODE REVIEW

**Comment by wenottingham** [comment]
Example of what happens on a celery/kombu app running in FIPS mode:

```
  File "/usr/lib/python2.7/site-packages/celery/app/control.py", line 80, in active
    return self._request('dump_active', safe=safe)
  File "/usr/lib/python2.7/site-packages/celery/app/control.py", line 70, in _request
    timeout=self.timeout, reply=True,
  File "/usr/lib/python2.7/site-packages/celery/app/control.py", line 306, in broadcast
    limit, callback, channel=channel,
  File "/usr/lib/python2.7/site-packages/kombu/pidbox.py", line 294, in _broadcast
    serializer=serializer)
  File "/usr/lib/python2.7/site-packages/kombu/pidbox.py", line 259, in _publish
    maybe_declare(self.reply_queue(channel))
  File "/usr/lib/python2.7/site-packages/kombu/utils/__init__.py", line 325, in __get__
    value = obj.__dict__[self.__name__] = self.__get(obj)
  File "/usr/lib/python2.7/site-packages/kombu/pidbox.py", line 224, in reply_queue
    return self.get_reply_queue()
  File "/usr/lib/python2.7/site-packages/kombu/pidbox.py", line 212, in get_reply_queue
    oid = self.oid
  File "/usr/lib/python2.7/site-packages/kombu/utils/__init__.py", line 325, in __get__
    value = obj.__dict__[self.__name__] = self.__get(obj)
  File "/usr/lib/python2.7/site-packages/kombu/pidbox.py", line 363, in oid
    oid = self._tls.OID = oid_from(self)
  File "/usr/lib/python2.7/site-packages/kombu/common.py", line 62, in oid_from
    return generate_oid(get_node_id(), os.getpid(), get_ident(), instance)
  File "/usr/lib/python2.7/site-packages/kombu/common.py", line 58, in generate_oid
    return str(uuid3(NAMESPACE_OID, ent))
  File "/usr/lib64/python2.7/uuid.py", line 528, in uuid3
    hash = md5(namespace.bytes + name).digest()
ValueError: error:060800A3:digital envelope routines:EVP_DigestInit_ex:disabled for fips
```

Note: I am a random product manager who knows just enough python but very little about kombu internals; this worked for the app we have, but may have consequences for the larger ecosystem I don't know about.

File: kombu/common.py [hunk 1]
```diff
@@ -50,7 +50,11 @@ def get_node_id():
 def generate_oid(node_id, process_id, thread_id, instance):
     ent = bytes_if_py2('%x-%x-%x-%x' % (
         node_id, process_id, thread_id, id(instance)))
-    return str(uuid3(NAMESPACE_OID, ent))
+    try:
+        ret = str(uuid3(NAMESPACE_OID, ent))
+    except:
```
Comment (georgepsarakis) [inline kombu/common.py]: Bare except statements should be avoided as they are too broad as they do not handle only the intended exception(s).

In your case it should be changed to:

```
except ValueError:
```

Comment (georgepsarakis) [inline kombu/common.py]: You could also add a test case patching the `uuid3` function to raise a `ValueError` exception always just to unit test this change.

**Comment by georgepsarakis** [comment]
@wenottingham let me know if you would like to apply the review suggestions. Otherwise I will merge this and I will open another Pull Request to include these.

**Comment by wenottingham** [comment]
Sorry about the delay. Fixed `except` clause, re-pushed.

I'll admit that I might be unlikely to get to the unit test.

**Comment by georgepsarakis** [comment]
@wenottingham thanks! Here is the test also (file `t/unit/test_common.py`):

```
def test_generate_oid():
    from uuid import NAMESPACE_OID
    from kombu.five import bytes_if_py2

    instance = Mock()

    args = (1, 1001, 2001, id(instance))
    ent = bytes_if_py2('%x-%x-%x-%x' % args)

    with patch('kombu.common.uuid3') as mock_uuid3, \
            patch('kombu.common.uuid5') as mock_uuid5:
        mock_uuid3.side_effect = ValueError
        mock_uuid3.return_value = 'uuid3-6ba7b812-9dad-11d1-80b4'
        mock_uuid5.return_value = 'uuid5-6ba7b812-9dad-11d1-80b4'
        oid = generate_oid(1, 1001, 2001, instance)
        mock_uuid5.assert_called_once_with(NAMESPACE_OID, ent)
        assert oid == 'uuid5-6ba7b812-9dad-11d1-80b4'
```

If you don't have time to add it, just let me know. I will merge this and open a separate Pull Request with the test.

**Comment by thedrow** [comment]
@georgepsarakis FYI there's no need for a separate PR. Github let's you push to his branches.

**Comment by georgepsarakis** [comment]
@wenottingham thanks, feel free to add yourself to the [AUTHORS](https://github.com/celery/kombu/blob/master/AUTHORS) file.

**Comment by elec3647** [comment]
Running into this issue with Red Hat Ansible Tower. Looking forward to seeing this fixed!

**Comment by auvipy** [comment]
isn't the fix already on pypi?

**Comment by georgepsarakis** [comment]
@elec3647 this should be included in versions 4.1.0 and onwards.

**Pull request merged**